# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/AudioConduit.cpp
# Commit: b5228352d20d
# Full Hash: b5228352d20db6c4899ddc3de268ebc35575f0d7
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 1
# Description:
#   Bug 1654112 - Fix up AudioConduit audio thread access model, and the channel count for silence. r=padenot
#   
#   This patch uses mMutex to guarantee that the audio thread does not get things
#   pulled away from under its feet. When on the AudioProxyThread we allow blocking
#   the thread during negotiation. On the real audio thread we avoid blocking on
# ==============================================================================

diff -r 056b43129502 -r b5228352d20d dom/media/webrtc/libwebrtcglue/AudioConduit.cpp
--- a/dom/media/webrtc/libwebrtcglue/AudioConduit.cpp	Fri Jan 29 15:02:58 2021 +0100
+++ b/dom/media/webrtc/libwebrtcglue/AudioConduit.cpp	Fri Jan 29 15:53:28 2021 +0100
@@ -471,7 +471,8 @@
     return kMediaConduitMalformedArgument;
   }
 
-  // if transmission is not started .. conduit cannot insert frames
+  // This is the AudioProxyThread, blocking it for a bit is fine.
+  MutexAutoLock lock(mMutex);
   if (!mSendStreamRunning) {
     CSFLogError(LOGTAG, "%s Engine not transmitting ", __FUNCTION__);
     return kMediaConduitSessionNotInited;
@@ -499,9 +500,16 @@
     return kMediaConduitMalformedArgument;
   }
 
+  // If the Mutex is taken, skip this chunk to avoid blocking the audio thread.
+  if (!mMutex.TryLock()) {
+    CSFLogError(LOGTAG, "%s Conduit going through negotiation ", __FUNCTION__);
+    return kMediaConduitPlayoutError;
+  }
+
   // Conduit should have reception enabled before we ask for decoded
   // samples
   if (!mRecvStreamRunning) {
+    mMutex.Unlock();
     CSFLogError(LOGTAG, "%s Engine not Receiving ", __FUNCTION__);
     return kMediaConduitSessionNotInited;
   }
@@ -518,6 +526,8 @@
     return kMediaConduitPlayoutError;
   }
 
+  mMutex.Unlock();
+
   CSFLogDebug(LOGTAG, "%s Got %zu channels of %zu samples", __FUNCTION__,
               frame->num_channels(), frame->samples_per_channel());
   return kMediaConduitNoError;